[FLINK-13581][coordination][tests] Harden BatchFineGrainedRecoveryITCase

azagrebin opened the pull request:
## What is the purpose of the change

If counting of mapper restarts in `BatchFineGrainedRecoveryITCase` is based on the open method of user function, the fact of the restart depends on internal implementation of the local Task and whether the open method is eventually called.

If execution attempt numbers are used instead, the test behaviour is more stable because it depends only on coordination. The execution attempt numbers can be queried from the REST client of the testing mini cluster.

## Brief change log

  - Introduce `MiniClusterClient` for `BatchFineGrainedRecoveryITCase` to query task attempts
  - Use mapper task attempt numbers instead of user function open method call counters

## Verifying this change

Run `BatchFineGrainedRecoveryITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java:
```diff
@@ -467,4 +509,66 @@ public void mapPartition(Iterable<Long> values, Collector<Long> out) throws Exce
 		}
 	}
+
+	private static class MiniClusterClient extends RestClusterClient<StandaloneClusterId> {
```
Why do we need to extend the `RestClusterClient`? Wouldn't it be enough to simply use the `RestClient` with the rest address which one gets from the `MiniCluster`? Then one would not have to change the access modifiers of the `RestClusterClient` and not have to jump through all these hoops to set up the client configuration.

azagrebin replied:
True, there is actually not so much code to reuse with the `RestClusterClient`.

azagrebin commented on flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java:
```diff
@@ -158,6 +195,10 @@ public void teardown() throws Exception {
+
+		if (client != null) {
+			client.shutdown();
+		}
```
`MiniClusterResource` provides `MiniCluster` with protected methods: `startTaskExecutor` and `terminateTaskExecutor` used in this test. `TestingMiniCluster` exposes them as public.

tillrohrmann replied:
All right. As a follow up one could think about providing a `TestingMiniClusterResource` but this is not necessary for this PR.

azagrebin commented on flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java:
```diff
@@ -110,13 +136,20 @@
+				1 + // this mapper
+				1 + // this mapper after PartitionNotFoundException
+				i + // previous mappers
+				i) // previous mappers after PartitionNotFoundException
```
Ok, we can try to simplify the test or you think this particular calculation could be rewritten?

tillrohrmann replied:
No I think the calculation itself with the comments is quite descriptive. However, the calculation itself is quite complicated because of all the different restarts being caused by different events.

azagrebin commented:
Thanks for the review @tillrohrmann and the feedback about the test.
Let's discuss later how to simplify the test in general.
I addressed the comments and left a question about the expected attempt number calculation.

tillrohrmann commented:
Hmm the japicmp plugin complains about the added methods to the `SupplierWithException` interface. I'll move them again to the Utils class.

tillrohrmann commented:
Actually, we don't need 0e09c48. Using `CompletableFuture#join` does the trick.
